chain/transaction.go

```diff
@@ -28,71 +28,82 @@ var (
 	_ mempool.Item = (*Transaction)(nil)
 )
 
-type Transaction struct {
-	Base *Base `json:"base"`
+type (
+	Actions     []Action
+	Transaction struct {
+		Base *Base `json:"base"`
 
-	Actions []Action `json:"actions"`
-	Auth    Auth     `json:"auth"`
+		Actions Actions `json:"actions"`
+		Auth    Auth    `json:"auth"`
 
-	digest    []byte
-	bytes     []byte
-	size      int
-	id        ids.ID
-	stateKeys state.Keys
-}
+		digest    []byte
+		bytes     []byte
+		size      int
+		id        ids.ID
+		stateKeys state.Keys
+	}
+)
 
-func NewTx(base *Base, actions []Action) *Transaction {
+func NewTx(base *Base, actions Actions) *Transaction {
 	return &Transaction{
 		Base:    base,
 		Actions: actions,
 	}
 }
 
+// Digest returns a byte slice containing the marshaled transaction without the auth.
+// method name might be misleading, as it returns the preimage of what typically
+// considered to be "the message digest"
 func (t *Transaction) Digest() ([]byte, error) {
 	if len(t.digest) > 0 {
 		return t.digest, nil
 	}
 	size := t.Base.Size() + consts.Uint8Len
-	for _, action := range t.Actions {
-		actionSize, err := GetSize(action)
-		if err != nil {
-			return nil, err
-		}
-		size += consts.ByteLen + actionSize
+
+	if actionsSize, err := t.Actions.Size(); err != nil {
+		return nil, err
+	} else {
+		size += actionsSize
 	}
+
 	p := codec.NewWriter(size, consts.NetworkSizeLimit)
-	t.Base.Marshal(p)
-	p.PackByte(uint8(len(t.Actions)))
-	for _, action := range t.Actions {
-		p.PackByte(action.GetTypeID())
-		err := marshalInto(action, p)
-		if err != nil {
-			return nil, err
-		}
+	if err := t.marshal(p, false); err != nil {
+		return nil, err
 	}
+
 	return p.Bytes(), p.Err()
 }
 
+// Sign sign the transaction using the provided AuthFactory. The pointer receiver is not being modified by this method,
+// and a new Transaction object would be returned. On success, the returned transaction is the signed transaction.
 func (t *Transaction) Sign(
 	factory AuthFactory,
 	actionRegistry ActionRegistry,
 	authRegistry AuthRegistry,
 ) (*Transaction, error) {
+	// grab the preimage of the transaction we would want to sign.
 	msg, err := t.Digest()
 	if err != nil {
 		return nil, err
 	}
+	// sign the preimage
 	auth, err := factory.Sign(msg)
 	if err != nil {
 		return nil, err
 	}
-	t.Auth = auth
```

Comment:
I figured it would be safer to create a new object for the marshaling rather than (set auth/marshal/clear auth).
i.e. the Sign method parameters suggest it would be a readonly method, as it returns the signed transaction.

Reply:
Yup, much better w/ this change